core/random.go

```diff
@@ -146,33 +145,32 @@ func (r *Random) GetLastRandomness(coinbase common.Address, db *ethdb.Database,
 		return common.Hash{}, nil
 	}
 
-	randomness := common.Hash{}
-	randomnessSlice, err := (*db).Get(commitmentDbLocation(lastCommitment))
+	parentBlockHashBytes, err := (*db).Get(commitmentDbLocation(lastCommitment))
 	if err != nil {
-		log.Error("Failed to get randomness from database", "commitment", lastCommitment.Hex(), "err", err)
-	} else {
-		randomness = common.BytesToHash(randomnessSlice)
+		log.Error("Failed to get last block proposed from database", "commitment", lastCommitment.Hex(), "err", err)
+		parentBlockHash := header.ParentHash
+		for {
+			blockHeader := chain.GetHeaderByHash(parentBlockHash)
+			parentBlockHash = blockHeader.ParentHash
+			if blockHeader.Coinbase == coinbase {
+				break
+			}
+		}
+		parentBlockHashBytes = parentBlockHash.Bytes()
 	}
-	return randomness, err
+	return crypto.Keccak256Hash(append(seed, parentBlockHashBytes...)), nil
 }
 
 // GenerateNewRandomnessAndCommitment generates a new random number and a corresponding commitment.
 // The random number is stored in the database, keyed by the corresponding commitment.
-func (r *Random) GenerateNewRandomnessAndCommitment(header *types.Header, state *state.StateDB, db *ethdb.Database) (common.Hash, error) {
+func (r *Random) GenerateNewRandomnessAndCommitment(header *types.Header, state *state.StateDB, db *ethdb.Database, seed []byte) (common.Hash, error) {
 	commitment := common.Hash{}
-
-	randomBytes := [32]byte{}
-	_, err := rand.Read(randomBytes[0:32])
-	if err != nil {
-		log.Error("Failed to generate randomness", "err", err)
-		return commitment, err
-	}
-	randomness := common.BytesToHash(randomBytes[:])
+	randomness := crypto.Keccak256Hash(append(seed, header.ParentHash.Bytes()...))
 	// TODO(asa): Make an issue to not have to do this via StaticCall
-	_, err = r.iEvmH.MakeStaticCall(*r.address(), computeCommitmentFuncABI, "computeCommitment", []interface{}{randomness}, &commitment, gasAmount, header, state)
-	err = (*db).Put(commitmentDbLocation(commitment), randomness[:])
+	_, err := r.iEvmH.MakeStaticCall(*r.address(), computeCommitmentFuncABI, "computeCommitment", []interface{}{randomness}, &commitment, gasAmount, header, state)
+	err = (*db).Put(commitmentDbLocation(commitment), header.ParentHash.Bytes())
 	if err != nil {
-		log.Error("Failed to save randomness to the database", "err", err)
+		log.Error("Failed to save last block proposed to the database", "err", err)
```

Comment:
nit: We're not saving the last block proposed here, are we? We're saving the parenthash of that block